lib/hooks/userspace/hook-grpc.js

```diff
@@ -19,14 +19,71 @@
 var cls = require('../../cls.js');
 var shimmer = require('shimmer');
 var semver = require('semver');
+var SpanData = require('../../span-data.js');
+
 var agent;
 
 var SUPPORTED_VERSIONS = '0.13 - 0.15';
 
-function startBatchWrap(startBatch) {
-  return function startBatchTrace(thing, callback) {
-    // TODO: maybe we only want to do this if a root context exists.
-    return startBatch.call(this, thing, cls.getNamespace().bind(callback));
+function clientMethodWrap(name, method, class_options) {
+  return function clientMethodTrace() {
+    var root = cls.getRootContext();
+    if (!root) {
+      agent.logger.debug('Untraced gRPC call: ', name);
+      return method.apply(this, arguments);
+    } else if (root === SpanData.nullSpan) {
+      return method.apply(this, arguments);
+    }
+    var span = agent.startSpan('grpc-call');
+    // Check if the response is through a stream or a callback.
+    if (!method.responseStream) {
+      // Grab the callback which is always required.
+      // Depending on whether deprecatedArgumentOrder is set, the position
+      // of the callback function differs.
+      // We need to wrap the callback with the context, to propagate it.
+      var cbIndex;
+      if (class_options && class_options.deprecatedArgumentOrder) {
+        cbIndex = method.requestStream ? 0 : 1;
+      } else {
+        cbIndex = arguments.length - 1;
+      }
+      if (cbIndex >= 0 && cbIndex < arguments.length) {
```

Comment:
Can this test ever fail? The above comment would suggest that the callback is always present somewhere in the arguments.